[WIP] Trace object paths in StructuralEqual

gbonik opened the pull request.

junrushao committed:
Initial commit

yelite committed:
WIP

yelite committed:
Add expr printing

yelite committed:
Implement CommReducer printing

yelite committed:
Add var printing

yelite committed:
Update based on review feedback

yelite committed:
WIP

yelite committed:
Op sugaring

yelite committed:
WIP

yelite committed:
Finish the free var visitor

yelite committed:
Add concise scoping helper

yelite committed:
PrintAttr partial

yelite committed:
WIP

yelite committed:
WIP

yelite committed:
Fix build problems

yelite committed:
Add easy test

yelite committed:
Finish the code for basic test case

yelite committed:
Add some documentations

yelite committed:
Fix formatting issue in doc printer

gbonik committed:
Trace object paths in StructuralEqual

sfvaroglu committed:
[ONNX] Add ReduceSum opset13 support (non-dynamic) (#11606)

* [ONNX] Add ReduceSum opset13 support (non-dynamic)

* Add check

* Add support for constant axis

* noop

* Rework logic

junrushao commented on include/tvm/node/object_path.h:
```diff
@@ -0,0 +1,214 @@
+  ObjectPath MissingArrayElement(size_t index);
+
+  static constexpr const char* _type_key = "ObjectPath";
+  static constexpr const uint32_t _type_child_slots = 3;
```
we usually leave it unspecified in case there is future expansion of subclasses

gbonik replied:
Done, thanks!

gbonik commented on src/node/structural_equal.cc:
```diff
@@ -137,6 +328,13 @@ class RemapVarSEqualHandler : public SEqualReducer::Handler {
       auto& entry = task_stack_.back();
 
+      if (entry.force_fail) {
+        if (first_mismatch != nullptr) {
```
Should probably call `CheckResult(false)` here, if we want to continue supporting `assert_mode`

gbonik commented on include/tvm/node/structural_equal.h:
```diff
@@ -89,6 +91,15 @@ class StructuralEqual : public BaseValueEqual {
+   * Same as operator(), but if `false` is returned, then `out_mismatch.lhs_path` and
+   * `out_mismatch.rhs_path` are set to the paths of the first mismatch.
+   */
+  bool FindFirstMismatch(const ObjectRef& lhs, const ObjectRef& rhs,
```
I haven't implemented this one yet, so the only entry point right now is from python

echuraev committed:
[OpenCL] Implement conv2d_winograd algorithm for Adreno (#11543)

* Implement conv2d_winograd algorithm for Adreno

* Implement gtest for OpenCL texture pool

* Implement conv2d_nhwc_winograd for Adreno

* Minor refactoring

* Fix lint

* Apply comments

* Apply comments

* Fix lint

junrushao commented on include/tvm/node/object_path.h:
```diff
@@ -0,0 +1,214 @@
+  virtual std::string LastNodeString() const = 0;
+
+ private:
+  ObjectPtr<ObjectPathNode> parent_;
```
just nitpicking on a convention in TVM: `ObjectPtr` is usually used as a temporary object returned from `make_object` or `CopyOnWrite`. As class members, usually we use `ObjectRef`. In this particular case, however, I'm not sure what the best approach could be to avoid cyclic dependency, but I myself usually go for `ObjectRef` as the most common type

junrushao replied:
alternatively we may put this field in subclasses and virtualize the GetParent method but those are just nitpicking and i dont have strong opinions

gbonik replied:
I actually started with the virtual GetParent(), but then storing the parent in the base class simplified the code and reduced the amount of boilerplate which I think is nice. Let me try the ObjectRef idea.

junrushao commented on include/tvm/node/object_path.h:
```diff
@@ -0,0 +1,214 @@
+
+// ----- Unknown attribute access -----
+
+class UnknownAttributeAccessPathNode final : public ObjectPathNode {
```
Just curious, when will we use this class?

gbonik replied:
This one is created by `ObjectPathNode::Attr()` when the passed `attr_key` is null. The idea is to keep going even if the failed to determine the attribute key from its address for some reason.

junrushao commented on include/tvm/node/structural_equal.h:
```diff
@@ -110,12 +124,24 @@ class SEqualReducer : public BaseValueEqual {
+    virtual bool SEqualReduce(const ObjectRef& lhs, const ObjectRef& rhs, bool map_free_vars,
+                              const ObjectPathPair& current_paths) = 0;
+
+    /*!
+     * \brief Mark the comparison as failed, but don't fail immediately.
+     *
+     * This is useful for producing better error messages when comparing containers.
+     * For example, if two array sizes mismatch, it's better to mark the comparison as failed
+     * but compare array elements anyway, so that we could find the true first mismatch.
+     */
+    virtual void DeferFail(const ObjectPathPair& mismatch_paths) = 0;
```
this is great idea btw

asparkhi committed:
[CMSIS-NN] Removed redudant arguments to CMSIS-NN wrapper function (#11431)

Removed input_scale and filter_scale from CMSIS-NN
wrapper function. These are not needed by CMSIS-NN
API which gets called from the generated C wrapper
function for Conv2D.

FranckQC committed:
[TIR] CSE pass : Restrict the equivalence to be decided by a normal form - avoids comparison of terms (#11574)

The CSE pass had been designed for potentially allowing comparisons (and commonings) of equivalent terms (like (x+y)+z and x+(y+z)), where **the notion of being equivalent was customizable, and no assumption was made about it**. That means that the implementation of the equivalence test function `EquivalentTerms()` - which was at the moment just calling the syntactical equality test `EqualTerms()` - could be replaced later by a cleverer equality test.

However, having such a generic way of comparing elements meant that in the function `SyntacticToSemanticComputations()`, where we were going from a hashtable of syntactical entities to what I called a vector of "semantical entites" (which are just canonical forms/representants of classes of equivalence of terms), **the only way was to compare each pair**.
That resulted in a quadratic behavior of this function, but there was no way around it as in order to merge equivalent entities into their class of equivalence, we had to compare them.

**This PR essentially does the following:**

- When computing the classes of equivalences of terms (therefore transforming a ComputationTable (i.e. a hashtable) into a vector of classes of equivalence) : **instead of comparing each pair of terms, relies on a normalization procedure to obtain a normal form for each of them**.
That transforms a small part of the algorithm that was quadratic to n.logn. However, it's difficult to see improvements in practice, in particular for average sized programs, as that part was a "small" quadratic to a "big" n.logn (finding things in a hash-table, copying it to a vector, etc).
It was probably going from a complexity of ~O(((n²-n)/2) + n.logn) to a complexity of ~O(3n + n.logn), so potential gains would only be expected for very large programs.

- Completely gives the user the possibility to turn ON/OFF the semantical comparisons of terms. It is turned OFF by default (as it's quite longer to compile with it ON, unsurprisingly), which means that by default, the equivalence coincides with the (syntactical) equality of terms.
    As the pass was written with the possibility to do these additional commonings (like (x+y)+z and x+(y+z)), it was a good time to fully plug that completely, up to the Python user who can now turn that ON if he wants to. But again, it is OFF by default, so no real change on that.

To run it ON, simply do:
`with tvm.transform.PassContext(config={'tir.enable_equiv_terms_in_cse_tir':True}):`
before calling `build()`

- When this boolean is set to ON, it uses a simple implementation of the normalization function with equivalences that uses `arith::Analyzer::Simplify` as noted by in https://github.com/apache/tvm/pull/10544 . Note that this is not a real normalization procedure as it is incomplete (i.e., it is not guarantee to converge to the normal form), but it is correct, and it works well with most properties : associativity of +, distributivity of * on +, etc.

- Clarifies and enhance the test base for the pass. In particular, it adds the tests that were written in https://github.com/apache/tvm/pull/10544 but which did not make it through.

- Also add the test ( https://github.com/AndrewZhaoLuo/TVM-Sandbox/blob/19284ddbd6bb28af61c0c2aa8bb334c5c53731a7/tir/test_inconsistent_tir_lowering.py#L1 ) demonstrating the (older) non-deterministic lowering and put it into a proper test, as I found it useful for making sure that this does not happen again. It has been copied from https://github.com/apache/tvm/pull/10663 and only slightly adapted (in particular for doing the comparison of hashes automatically instead of printing them and relying on a human to compare them).

driazati committed:
[ci] Add guards to pytest_wrapper (#11553)

This should fix #11544 and adds some more logging in case the issue persists. Unfortunately it is difficult to test for real since the case data in that PR is thrown away after Jenkins is done (Jenkins does store test data but it marshals JUnits into its own format)

Co-authored-by: driazati <driazati@users.noreply.github.com>

sunggg committed:
[PASS] Refactor a couple of TIR passes - BindTarget, AnnotateEntryFunc, Filter, LowerInitBlock (#11628)

This PR fixes a few inconsistent pass registration and add testcases for them. 
- `LowerInitBlock` had mismatch between its pass name and ffi key.
- `BindTarget`, `AnnotateEntryFunc`, `Filter` were not following the name convention of tir passes and they were not registered in FFI registry.

guberti committed:
[microTVM] Remove microTVM RVM version suffix (#11629)

czh978 committed:
[Frontend][TFLite] Improve support for half_pixel_centers in resize (#11521)

* add resize_nearest_neighbor op test

* Improve support for half_pixel_centers in resize

asparkhi committed:
Making CMSIS-NN tests pylint compliant (#11625)

Lunderberg committed:
[TIR][Schedule] Allow named block and buffer arguments in Schedule (#11624)

* [Schedule] Allowed string argument as block arg

This has previously been implemented for `Schedule.transform_layout`
in https://github.com/apache/tvm/pull/11296, extending to allow for
block arguments in all `Schedule` methods.

This change was only made for arguments that must be a `BlockRV`.  For
arguments that may be either a `BlockRV` or another
type (e.g. `Schedule.get_child_blocks` accepts either `BlockRV` or
`LoopRV`), this sugar is not implemented, to avoid ambiguity.

* [Schedule] Allowed string argument to Schedule.reindex

Similar to https://github.com/apache/tvm/pull/11269, which added this
functionality to `Schedule.transform_layout`.

* CI test update

driazati committed:
[ci] Rebuild Docker images if necessary (#11329)

This rebuilds Docker images and uses them in later stages in the same build. If the build is running on `main`, then the images are uploaded to Docker Hub automatically once the run is complete. Images are always rebuilt, but Docker Hub functions as a cache. If there have been no changes to `docker/` since the last available hash on Docker Hub, then the build will just use the images from Hub.

Commit:
[Hexagon] Fix gtest flag in apps/hexagon_api/CMakeLists.txt (#11652)

mehrdadh committed:
[microTVM] Update pyproject to python3.7 (#11634)

* Update to python3.7 and add poetry.lock file

guberti committed:
[microTVM] Add support for Arduino Portenta H7 (#11636)

* Add support for Portenta H7

* Add Portenta H7 to supported boards in README

* Rerun tests

Commit:
adding vvchernov to contributors file (#11649)

Commit:
[COMMUNITY] Alexander Peskov -> Reviewers (#11648)

* adding ramana to reviewers list

* adding apeskov as reviewer

* fix

Commit:
[TVMSCRIPT] Improve tvmscript type hints (#11654)

* [TVMSCRIPT] Improve tvmscript type hints

- Change numeric types to classes so they work as function arguments.
- Add var as a class.
- Add floordiv, index, and mod to PrimExpr.

* use Union

fPecc committed:
[CRT runtime] Added functions TVMPlatformBeforeMeasurement and TVMPlatformAfterMeasurement (#11244)

* Added functions with weak links before and after TVMFuncCall in the TimeEvaluator

* Fixed lint

* Clang changes

* Added clang proposal

* clang-format proposed changes

Co-authored-by: Federico Peccia <peccia@fzi.de>

mbs-octoml committed:
[BUG] Disable second PlanDevices pass (#11662)

Though started with the best of intentions, the second
PlanDevices pass to account for memory scope's introduced
by lowering is buggy and not ready for prime time. It
has caused an ICHECK fail since for some reason the new
constraints are not flowing into device_copies.

masahi committed:
[TIR, CUDA] Add pass to replace global to shared memory copy with cp.async (#11658)

* [TIR, CUDA] Add pass to replace global to shared memory copy with cp.async

* add missing doc

* black

* missing src

* clang format

* clang format

* check against nested async scope

mbs-octoml committed:
[BYOC] RelayToTIR custom codegen passes can still depend on dynamic shape functions (#11619)

In #11474 I got ready to switch CUTLASS from function-at-a-time to IRModule-at-a-time compilation.
However my approach didn't handle dynamic shape functions, so I adjust it here.

The idea is still that such passes will leave behind
calls to 'extern' functions. However, converting those
calls to 'call_lowered' form in
MarkCompilerFunctionsAsExtern is too soon since only
the TECompiler knows how to capture all the attributes
necessary to support dynamic shape functions.

So stop doing that in MarkCompilerFunctionsAsExtern and
instead support this case properly in the TECompiler.

While there try to chip away at the chronic lack of structure in te_compiler.cc. Every little bit helps.

Add a basic unit test.

junrushao committed:
[MetaSchedule] Developer Ergonomics Enhancement (#11622)

Per discussion with @Kathryn-cat

- [x] Move `initialize_with_tune_context` as private API `_initialize_with_tune_context`, and
encourage using `TuneContext.initialize`
- [x] Instead of using bunch of import statements, encourage using `ms.xxx` as the prefix
(e.g. `ms.database.MemoryDatabase`) to organize things better
- [x] Move `DefaultLLVM`, `DefaultCUDA` to a separate file and make them more discoverable
- [x] Move `DummyDatabase` to `tvm.meta_schedule.database.MemoryDatabase` given it's actually useful
- [x] Delegate class members' methods in `TuneContext`, for example, having
`TuneContext.generste_design_space` from `TuneContext.space_generator.generste_design_space`

Next PR:
- Allow using a string `"default"` in `TuneContext` as well as `tune_relay/tir/te` to quickly
specify a set of target-specific rules
- Add `TuneContext.tune` to allow directly tuning without task scheduler.
- Enhance detection of `ScheduleFn` in `TuneContext` to make it easier for users to quickly try out
template-driven scheduling on TIR.

Co-Authored-By: Kathryn (Jinqi) Chen <65606304+Kathryn-cat@users.noreply.github.com>

yuanfz98 committed:
[Bugfix] GetReduceAxes accept empty axis (#11643)

* emptycommit 2nd try

* code

Co-authored-by: yuanfz <42092999+FZYUAN-1@users.noreply.github.com>

Qianshui-Jiang committed:
[DNNL][Relay extern-schedule] DNNL Conv2D Kernel enable by assigning "-libs=mkldnn" (#11571)

* enable oneDNN conv op by using -libs=mkldnn

* add channel last format support and let oneDNN chose blocked format.

* remove unnecessary changes

* reformat 3 files

* reformat 1 file

* change the argument name

* change the argument name

* rename the arguments

* fix cpp lint issue

* fix cpp lint issue

* fix cpp lint issue

* clang reformated

* adjust .py import for testing

* function existence check in test

NicolaLancellotti committed:
Add assert message (#11665)

Change-Id: I88f19c7105cce048d2f52d50450a551fb12162dc

gigiblender committed:
[CI] fix ci_gpu dockerfile (#11644)

mehrdadh committed:
[microtvm] Add mxnet importer and update pyyaml to fix poetry error (#11668)

driazati committed:
[ci][docs] Don't delete old versions when checking out docs (#11612)

We don't have a good way to tell if a file was deleted or not in a docs update, so currently we delete the entire `docs/` folder and replace it from the build. However, this includes old version docs that aren't build in the normal docs build. This excludes them from the deletion so they stick around between updates. We'll have to revisit this list at each release but it should be a simple update.

Co-authored-by: driazati <driazati@users.noreply.github.com>

mehrdadh committed:
[Hexagon] Run single RPC server on Android in each testing session  (#11547)

* Reuse hexagon launcher in test session

* separate random name generation

* revert get_aot_executor

* Fix launcher for simulator case

* add stop server for simulator

billishyahao committed:
[BYOC][DNNL] Improve performance of DNNL BYOC dense operator (#11513)

* Enhance dnnl byoc dense operators performance by 1) introducing gelu fusion and 2) introducing alter dense weight layout.

* fix lint issue

* add unittest for dense pack

* Make code compatible after introducing TensorRequisite(PR-11345)

* Fix comments & refactor code

* Fix lint

* Fix partition graph unittest case

* Fix comments

* Fix comments

* Fix lint

Qianshui-Jiang committed:
[DNNL][CBLAS][BYOC] Unifles all MKLDNN/DNNL  to DNNL (#11638)

* unifies all MKLDNN/DNNL_CODEGEN to DNNL

* translate -lib=mkldnn to -libs=dnnl in target

* type check added before

* rebase and update conv2d from mkldnn to dnnl

Commit:
[FIX,METASCHEDULER] Fix tune_te (#11676)

`tune_te` was broken because it passed a primfunc to `tune_tir`. Now it
is wrapped in an IRModule. Also the test is re-enabled.

Kathryn-cat committed:
[MetaSchedule] Generate MetaSchedule Dataset (#11641)

In order to build a dataset for improving the cost model for MetaSchedule, I added several files
including importing models to TVM, extracting tuning tasks, and sampling measure candidates.
Meanwhile, I exposed some methods in C++ to the Python side to assist the process.

mbs-octoml committed:
[BYOC] Make CUTLASS BYOC integration 'Collage friendly'   (#11631)

* [BYOC] Make CUTLASS BYOC integration 'Collage friendly'

(See https://discuss.tvm.apache.org/t/byoc-supporting-cutlass-byoc-with-collage/12796/6 for
context, which in turn is part of Collage (https://github.com/apache/tvm-rfcs/blob/main/rfcs/0062-collage.md).

Currently CUTLASS has four entry points:
 - The usual 'partition_for_cutlass' partitioning function, using the
   standard pattern table and pass machinery (see cutlass/build.py).
 - A 'tune_cutlass_kernels' function which augments CUTLASS partition
   functions with the results of building and running test kernels (see cutlass/build.py).
 - A 'relay.ext.cutlass' external codegen function which inspects the
   turning results and generates a CSourceModule for each partitions
   (see cutlass/codegen.cc).
 - A 'build_cutlass_kernels_vm' function which runs 'export_library' with
   all the nvcc compiler options needed to build all the CSourceModules
   (see cutlass/bild.py).

For Collage we'd like CUTLASS to have only two entry points: 'partition_for_cutlass',
and 'relay.ext.cutlass' or equivalent. This makes the CUTLASS external codegen integration
composable with other integrations, which in turn helps Collage avoid having to understand any
external codegen APIs other than the global pattern table and the custom compilation function/pass.

Collage also tends to end up requiring multiple partitions for the same backend since it is
more aggressive at mixing-and-matching smaller sub-graphs between backends. Thus we'd also like
to make sure all tuning, generated code and compilation overhead is shared between all such CUTLASS
partitions.

So, in this PR:
 - We add all the CUTLASS-specific tuning and compilation options as new Target
   attributes for the 'external codegen' "cutlass" TargetKind (cutlass/target.cc).
   The user now has one place to provide those settings, and we've already done the
   legwork to plumb the target instance.
 - We replace 'relay.ext.cutlass' with a 'RelayToTIR' custom pass hook
   'CompileForCutlass' (see cutlass/codegen.cc). This pass obviously can see all
   the CUTLASS partitions in the IRModule, so we can now share tuning results
   between them all and can be sure to generate a single CSourceModule. The pass can
   also invoke the compiler to yield a StaticModule, which we've also already done the
   legwork to support. In this way all CUTLASS-specific steps are handled at once.
 - For convenience we supply 'finalize_modules' and 'finalize_modules_vm' which
   invoke nvcc for final linking (using export_library as usual). However, there's now
   nothing CUTLASS specific in those helpers other than their overriding of the 'compiler' to
   be nvcc.
 - test_cutlass.py is updated to use the new API.

 Though this is a breaking change for existing users of the CUTLASS integration the
 change is pretty minor, as shown in test_cutlass.py.

* - Masa's comments

* - Remove unnecessary save.

vinx13 committed:
[TIR] Register CUDA WMMA tensor intrinsics (#11677)

* Register CUDA wmma tensor intrins

* Meta programming to generate wmma intrin

* format

* fix

* fix wmma_store

* lint

* Update cuda.py

billishyahao committed:
Fix typos in target warn of dnnl (#11678)

junrushao committed:
[MetaSchedule] JSONDatabase Utilities (#11680)

This PR adds some utility to JSONDatabase to accelerate its loading/saving time.

mbs-octoml committed:
[Relay] Finish implementations of WithFields (#11674)

wzh99 committed:
[Bugfix] Shape inference of weight for grouped `nn.conv3d` (#11681)

* Fix `nn.conv3d` weight shape inference.

* Add test for conv3d type inference with groups.

masahi committed:
fixed cutlass byoc build break (#11686)

driazati committed:
[docs] Add links to v0.8.0 docs (#11647)

This uses the new code from https://github.com/tlc-pack/tlcpack-sphinx-addon/pull/5 with a link to the v0.8.0 docs. We can update this in the future as we add more releases.

Co-authored-by: driazati <driazati@users.noreply.github.com>

Lunderberg committed:
[UnitTests] Parametrized test_topi_argwhere.py (#11651)

Refactored while debugging breakage of tests in
https://github.com/apache/tvm/pull/11646.  Submitting as a separate
PR, as it isn't necessary or related to the primary changes in that
PR.

asparkhi committed:
Added a docstring to missing CMSIS-NN test (#11690)

* Made CMSIS-NN tests pylint compliant

Change-Id: I6bc536a80a24a1603e9f75f8ee9a26d0d88f10df

* Removed comments that disabled pylint checks

Change-Id: Iee513a4a5bef1db5b78e1d25a30ac7202f8b0e92

* Fixed pylint issue in the generate_constants test

Change-Id: Icd341cf524b331ced1fc7ef282b67296583b0fa4

csullivan committed:
[Hexagon] Tighten requirements on inclusion of runtime sources (#11635)

* Tighten requirements on when Hexagon runtime sources
are included in the runtime build. Specifically only include them
when building for hexagon rpc on hardware and do not include them
for x86 (host, simulator) or android builds.

* Remove device_api.cpu binding to hexagon in simulator rpc session.

Co-authored-by: Adam Straw <astraw@octoml.ai>
Co-authored-by: Karl Koscher <kkoscher@octoml.ai>

* if(BUILD_FOR_HEXAGON)

Co-authored-by: Adam Straw <astraw@octoml.ai>
Co-authored-by: Karl Koscher <kkoscher@octoml.ai>

zxybazh committed:
[MetaSchedule] Add Profiler Support For Tuning Efficiency Optimization (#11486)

Co-authored-by: Junru Shao <junrushao1994@gmail.com>

csullivan committed:
[TE] Support schedulable TIR compute definitions in TOPI (#11589)

This PR adds `te.extern_primfunc` which provides the interface around TE ExternOp that allows a TVMScript defined schedulable TIR PrimFunc to be inlined into a TE compute graph. The result is that TIR can be used for compute definitions in Relay OpStrategies and, paired with meta-scheduler support in relay as introduced in #10578, these compute definitions can be scheduled and tuned as demonstrated in the attached tests.  

Prior to this, compute definitions were limited to those definable in TE only. As a consequence of this patch and ongoing improvements to TVMScript meta-programming (#11097), TOPI can be extended to include compute and scheduling functions targeting schedulable TIR uniformly.

nverke committed:
Updated install from source docs to include additional instructions for M1 macs. (#11675)

Co-authored-by: Noah Verke <noahverke@nverke-MBP.local>

lazycal committed:
Fix onnx round import with float64 inputs. (#11685)

* Fix onnx round import with float64 inputs.

* Fix lint and optimize dtype mapping.

adstraw committed:
[Hexagon] Add HexagonThreadManager (#11653)

* Adding initial threadmanager class

* Fixed compile errors

* Moving constant defines inside class

* Updating qurt includes

* use default scope for hexagon buffers

* Updating buffer allocations

* Fixed bug where array of pointers treated as array of structs

* - Updated HexgonDeviceAPI to use HexagonThreadManager
- Updated HexagonThreadManager interface to use TVMStreams
- Added second `Dispatch` interfce in thread manager to use PackedFuncs
- Updated thread manager to use vector for dynamic semaphore allocation
- Added "#if defined(__hexagon__)" in several places to prevent compilation errors

* Bug fixes + interface addition + basic thread tests
 - Fixed GetStreams not returning the streams properly
 - Added missing semaphore cleanup to prevent qurt kernel resource leakage
 - new interface functions:
   - Start() : now all worker threads are blocked on initialization until ThreadManager->Start() is called
   - WaitOnThreads() : blocking call which waits until all worker thread queues are empty
 - added extra debug logging
 - Two new basic thread tests working

* Adding initial ThreadManager tests

* HexagonThreadManager tests and refactor

* remove stack / pipe size member vars

* init pointers in the header file

* move all mem allocs to SpawnThreads

* start_semaphore as object instead of pointer

* fix bug with WaitOnThreads deadlock + Wait/Signal off by one error

* add / refactor Signal / Wait tests

* add SyncFromTo test cases

* add Dispatch test cases

* add pipe fill and overflow cases

* Updating dispatch to return bool and fix pipe overflow problem

* change around min / max values for stack / pipe

* integrate pipe fill / overflow tests back into HTM test suite

* use HexagonBuffer

* assert if stack / pipe sizes fall below min

* Changed semaphore vector to store pointers, not structs (fixes vector capacity adjustment invaliding in-use addresses).

* add producer consumer, thread order test cases

* change to unordered_map for semaphores and remove PreallocateSyncs

* tests running on device

* code cleanup for compile warnings

* remove #if defined(__hexagon__) guards

* copyright, format, lint

* add hexagon buffer map class

* remove redundant thread manager tests

* revert Hex Dev API changes for threading

* add comments; remove untested code to dispatch / wrap a packed func

* pass pipe address and not HTM pointer to thread context

* rename to HexagonBufferManager

* cleanup ahead of PR

* use DLOG(INFO)

* refactor GetStreamHandles to return a vector by value

* adjust HexagonBufferManager methods; use thread_manager file names

* style guidelines and debug prints

* reinterpret cast for TVMStreamHandle

* end member variables with underscore

Co-authored-by: Joseph McMahan <jmcmahan@octoml.ai>

AndrewZhaoLuo committed:
[AutoTVM][Autoscheduler] Default build funcs inherit PassContext (#11632)

* init commit

* lint

* empty commit

* test results

* reset progress

* lint

* fix

guberti committed:
[WIP] [CI] Bump CI GPU image version (#11637)

* Bump CI GPU image version

* Run generate,py

junrushao committed:
[Hotfix][MetaSchedule] Importing from test foldeer (#11695)

A concurrent merge breaks the unittest which imports directly from
`meta_schedule.testing`.

srkreddy1238 committed:
[BYOC-OpenCLML] OpenCLML integration with TVM. (#10243)

* [BYOC-OpenCLML] OpenCLML integration with TVM.

* [BYOC-OpenCLML] Cleanup and review.

mehrdadh committed:
cleanup (#11659)

Commit:
[tests][hexagon] Fix `allocate_hexagon_array` bug. (#11709)

Fix bug where `allocate_hexagon_array` in
`tests/python/contrib/test_hexagon/infrastructure.py` wasn't
respecting the caller-specified `memory_scope`.

Icemist committed:
Move FlattenAtrousConv before AlterOpLayout in the default opt pipeline. (#11706)

Co-authored-by: Andrey Malyshev <elvin.nnov@gmail.com>

Co-authored-by: Andrey Malyshev <elvin.nnov@gmail.com>

adstraw committed:
[Hexagon] remove #if defined(__hexagon__) where it is no longer needed (#11708)

* remove #if defined(__hexagon__) where it is no longer needed

* format and lint

quic-sanirudh committed:
[CI] [Hexagon] Update docker tag in jenkins (#11588)

alanmacd committed:
[microTVM][zephyr] Add support for host-driven AoT execution on zephyr (#11650)

* - add support for host-driven AoT execution on zephyr;
- add initial version of reference counting to prevent python code from inadvertently freeing tensors during garbage collection;
- add support for numerical indices to host-drive AoT get_input();
- add two initial tests for host-driven AoT execution on zephyr;
- rename existing zephyr AoT exec. test;

* address PR feedback

* increase stack size to accommodate qemu_riscv64 stack usage

Commit:
[LLVM] Update uses of AllocaInst::getAlign[ment] (#11718)

Today's LLVM main branch removed AllocaInst::getAlignment.

lazycal committed:
Fix 1d-softmax schedule. (#11719)

junrushao committed:
[MetaSchedule] Apply-History-Best Task Filtering (#11692)

This PR enables task filtering in Apply-History-Best, which is used in
Relay/Relax integration. Previously, even though a task is ruled out
during task extraction, it still shows up in Relay compilation due to
the lack of filtering on `Apply-History-Best`. However, TE-to-TIR
conversion `te.CreatePrimFunc` doesn't support all cases with hybrid
operators involved, which leads to post-tuning failure affecting
multiple models.

junrushao committed:
[Bugfix][TIR] Narrow-Datatype for thread axis (#11725)

This PR fixes a bug in the pass Narrow-Datatype in TIR, where dtype of
certain IterVar and loop variables are adjusted to narrower ones.

The bug occurs when the dtype of thread axis is int32, while its extent
is int64, where the original behavior will not narrow the extent to
int32, which causes an assertion thrown in IterVar's constructor. An
alternative approach is to re-dtype IterVar to int64, however, the
subsequent passes do not actually respect int64 thread axes, which leads
to even more issues in lowering.

This bug prevents AutoTIR in tuning Huggingface DistilBERT.

gigiblender committed:
[CI] add GH workflow to comment with link to docs (#11594)

Mousius committed:
[CI] Apply linting rules to AOT tests (#11657)

This enables pylint against the AOT test cases.

One issue I found was with the `tvm.testing.parameter` which breaks the naming convention rules in pylint (constants are upper case and function parameters are lower case). It may be worth a syntax similar to:

```
tvm.testing.parameter("enable_usmp", [True, False])
```

asparkhi committed:
[CMSIS-NN] Fixed error in finding input's dtype in maxpool (#11701)

jiangjiajun committed:
[TVMC] Fix error while compile paddle model with tvmc (#11730)

The tvmc command will throw a error while the passed path of model is not exist, But for PaddlePaddle model, it contains 2 file model_name.pdmodel and model_name.pdiparams, we only pass the prefix like inference_model/model_name.

This pr is same with https://github.com/apache/tvm/pull/11108 
Since the origin PR didn't update for a long time, I send this new PR

Commit:
[MetaSchedule] Include te/tensor.h instead of forward declaring te::Tensor (#11731)

ApplyHistoryBestNode declares an Array of Tensor. There are type traits
used in Array that require that the element type is complete at the time
of the declaration. With only a forward declaration compilation fails
(clang 14.0.3, libc++).

jverma-quic committed:
[Hexagon] Implement avg_pool2d slice op (#11417)

* Implement avg_pool2d slice op

* Address review comments and fix the STIR schedule

* Fix formatting issues

* Address pylint errors

* Additional formatting issues

* more pylint fixes

* Changed arch version to v68 for now

* Changing arch version back to v69

* Move the test to tests/python/contrib/test_hexagon/topi

gigiblender committed:
[ci] Skip failing tests in wheel (#11705)

Some python tests are failing in the wheel. This PR skips them if the environment variable `WHEEL_TEST` is set.

This PR is related to https://github.com/tlc-pack/tlcpack/pull/115.

junrushao committed:
[MetaSchedule] Developer Ergonomics Enhancement II (#11727)

Follow-up of #11622, per discussion with @Kathryn-cat

- [x] Allow using a string `"default"` in `TuneContext` to quickly specify a set of target-specific
rules
- [x] Enhance detection of `ScheduleFn` in `TuneContext` to make it easier for users to quickly try
out template-driven scheduling on TIR.

Next PR:
- Add `TuneContext.tune` to allow directly tuning without task scheduler.

Co-Authored-By: Kathryn (Jinqi) Chen <65606304+Kathryn-cat@users.noreply.github.com>

alanmacd committed:
rename aot_demo to aot_standalone_demo for clarity vs. host-driven aot (#11723)

zxybazh committed:
[MetaSchedule] Modify Profiler Timers (#11735)

Minor modification to scoped timers to cover 99% of all the time cost during MS tuning. Allow `ApplyHistoryBest` and `TaskExtraction` time to be counted during tune_relay.

guberti committed:
[microTVM] Add support for the Raspberry Pi Pico via Arduino (#11694)

* Add RP2040 support

guberti committed:
[microTVM] [docs] Point micro_train tutorial links to official repos (#11715)

* Point micro_train tutorial links to official repos

yuanfz98 committed:
[Pytorch] Add quantized::leaky_relu (#11729)

* emptycommit 2nd try

* add operator and test

* example output

* lint with black

* register param index

* remove assert as it is a warning in torch

* fix algo bug

Co-authored-by: yuanfz <42092999+FZYUAN-1@users.noreply.github.com>

junrushao committed:
[TIR] Add preserve-unit-iters (#11585)

wrongtest-intellif committed:
[TVMScript] Support roundtrip of LetNode (#11742)

Just a missing support for `tir.LetNode`

wzh99 committed:
[Relay] Implement `SoftmaxRel` for softmax operators. (#11728)

* Implement `SoftmaxRel` for softmax operators.

* Print better error message for wrong axis.

junrushao committed:
[Bugfix][MetaSchedule] Filter out dynamic extents (#11747)

Previously only static shape computation is allowed in our tuning
system. However, one special case is overlooked: the reduction iter vars
could still have dynamic iteration domains which depend on other data
parallel vars. This PR rules out this case by carefully checking all the
loop extents during task extraction.

Related issue: https://github.com/apache/tvm/issues/11746.

gigiblender committed:
[ci] Remove apt cache from the docker images (#11470)

mehrdadh committed:
[microTVM] Refactor RVM scripts and fix DNS network issue (#11741)

* refactor scripts

* address comments

AndrewZhaoLuo committed:
upgrade ci lint docker file (#11734)

areusch committed:
Fix CI break due to concurrent merge. (#11753)

* #11470 and #11741.

mehrdadh committed:
[skip ci][microTVM] Update Arduino RVM name and box version (#11743)

* update

* Fix version

* readme

* Update README.md

mehrdadh committed:
[skip ci][microTVM] Update Zephyr RVM name and box version (#11655)

* Use new Zephyr RVM version

* fix box name

* fix version to match zephyr

* update version

* Update README.md

* Update README.md

junrushao committed:
[MetaSchedule][Minor] Organize Testing Scripts (#11751)

supersat committed:
Add optional mem_scope parameter to tvm.nd.array and tvm.nd.copyto (#11717)

huajsj committed:
[ci][docker gpu] Install dnnl in docker GPU. (#11744)

BYOC related tutorial may use dnnl  and such tutorial run at docker gpu
which need to install dnnl to prepare the environment.

anonymous-me-prog committed:
Constant name prefix added (#11509)

This is a proposal to fix the bug reported here: https://discuss.tvm.apache.org/t/problem-with-allocateconstnodes-in-cmsis-nn-code/12806

Bug report: #11394

A prefix has been added to the "constant_" in te_compiler_cache.cc to distinguish from the constant naming generated in aot_executor_gen.cc

apeskov committed:
Enable QNN primitives for DNNL runtime (#11642)

* [DNNL] Enable QNN primitives

Signed-off-by: Alexander Peskov <peskovnn@gmail.com>

* [DNNL] add qnn test

Signed-off-by: Alexander Peskov <peskovnn@gmail.com>

* typo fix

Signed-off-by: Alexander Peskov <peskovnn@gmail.com>

Commit:
[hexagon][testing] add test-skip logic; fixes (#11737)

- Skip Hexagon benchmarks whenever the env. var `ANDROID_SERIAL_NUMBER`
  has the value `simulator`.

  This is a temporary hack to prevent the CI pre-commit hook from
  running benchmarks, due to the extra time required.

- Fix a bug where the elementwise-add benchmark code was broken by
  an earlier change to the `HexagonLauncherRPC` class.

- Rename `benchmark_elemwise_add.py` to `test_benchmark_elemwise_add.py`
  so that it's noticed by the CI test infrastructure.
  (CI tests are sometimes run in contexts _other than_ the pre-commit
  hook.)

- Miscellaneous small changes to
  `tests/python/contrib/test_hexagon/benchmark_util.py`.

asparkhi committed:
[CMSIS-NN] Fixed the case with repeating operands in the QNN binary ops (#11732)

Kathryn-cat committed:
[MetaSchedule] Distributed Measurement (#11683)

This PR includes the distributed measurement of tuning candidates using builder and async runner, as well as some auxiliary functions. It enables multiple builders and multiple runners with a tracker connecting in between. The hierarchy of files in the database can be further compacted to make the database more concise.

Commit:
[hexagon][testing] add max_pool2d benchmark (#11720)

- Add benchmarking framework for Hexagon maxpool-2d kernels,
  and one (simple) kernel.

mehrdadh committed:
[MLF] Add support for multiple modules in Model Library Format (#11464)

Raghav-Chakravarthy committed:
[Runtime][PipelineExecutor] Added Interface to Track Number of Global Inputs (#11315)

* [Runtime][PipleineExecutor] Added Interface to Track Number of Global Inputs

Added a feature to PipelineExecutor to track number of Global Inputs.

* Fixed CI Error

* Fixed remaining CI Error

zxybazh committed:
[MetaSchedule][Minor] Fix EvaluatorConfig Argument Description (#11766)

Pointed out by @sunggg that the description of `number` and `repeat` for evaluator configuration is not accurate, updated to a version more consistent with `TimeEvaluator`.

![TimeEvaluator](https://user-images.githubusercontent.com/3203174/174385966-74d3dbf6-dcca-43ea-9c0b-a91b4a281687.png)

areusch committed:
Add tool to clear stale images. (#11772)

zxybazh committed:
[MetaSchedule][Minor] Add Describe Function For Tuning Scripts (#11754)

This PR is based on #11751 and adds `describe` function for `tune_relay` and `tune_onnx` script on both AutoScheduler and MetaSchedule. It prints out very useful information for reproducibility as follows:
```
Python Environment
  TVM version    = 0.9.dev0
  Python version = 3.8.8 (default, Apr 13 2021, 19:58:26)  [GCC 7.3.0] (64 bit)
  os.uname()     = Linux 5.15.5-76051505-generic #202111250933~1638201579~21.04~09f1aa7-Ubuntu SMP Tue Nov 30 02: x86_64
CMake Options:
  {
    "BUILD_STATIC_RUNTIME": "OFF",
    "COMPILER_RT_PATH": "3rdparty/compiler-rt",
    "CUDA_VERSION": "NOT-FOUND",
    "DLPACK_PATH": "3rdparty/dlpack/include",
    "DMLC_PATH": "3rdparty/dmlc-core/include",
    "GIT_COMMIT_HASH": "3b872a0adae07b0cd60248346fd31b158cba630c",
    "GIT_COMMIT_TIME": "2022-06-15 11:27:59 -0700",
    "HIDE_PRIVATE_SYMBOLS": "OFF",
    "INDEX_DEFAULT_I64": "ON",
    "INSTALL_DEV": "OFF",
    "LLVM_VERSION": "11.0.1",
    "PICOJSON_PATH": "3rdparty/picojson",
    "RANG_PATH": "3rdparty/rang/include",
    "ROCM_PATH": "/opt/rocm",
    "SUMMARIZE": "OFF",
    "TVM_CXX_COMPILER_PATH": "/usr/lib/ccache/c++",
    "USE_ALTERNATIVE_LINKER": "AUTO",
    "USE_AOT_EXECUTOR": "ON",
    "USE_ARM_COMPUTE_LIB": "OFF",
    "USE_ARM_COMPUTE_LIB_GRAPH_EXECUTOR": "OFF",
    "USE_BLAS": "none",
    "USE_BNNS": "OFF",
    "USE_BYODT_POSIT": "OFF",
    "USE_CLML": "OFF",
    "USE_CLML_GRAPH_EXECUTOR": "OFF",
    "USE_CMSISNN": "OFF",
    "USE_COREML": "OFF",
    "USE_CPP_RPC": "OFF",
    "USE_CUBLAS": "OFF",
    "USE_CUDA": "/usr/lib/cuda-11.2",
    "USE_CUDNN": "OFF",
    "USE_CUSTOM_LOGGING": "OFF",
    "USE_CUTLASS": "OFF",
    "USE_DNNL": "OFF",
    "USE_ETHOSN": "OFF",
    "USE_FALLBACK_STL_MAP": "OFF",
    "USE_GRAPH_EXECUTOR": "ON",
    "USE_GRAPH_EXECUTOR_CUDA_GRAPH": "OFF",
    "USE_GTEST": "AUTO",
    "USE_HEXAGON": "OFF",
    "USE_HEXAGON_GTEST": "/path/to/hexagon/gtest",
    "USE_HEXAGON_RPC": "OFF",
    "USE_HEXAGON_SDK": "/path/to/sdk",
    "USE_IOS_RPC": "OFF",
    "USE_KHRONOS_SPIRV": "OFF",
    "USE_LIBBACKTRACE": "ON",
    "USE_LIBTORCH": "OFF",
    "USE_LLVM": "llvm-config-11",
    "USE_METAL": "OFF",
    "USE_MICRO": "OFF",
    "USE_MICRO_STANDALONE_RUNTIME": "OFF",
    "USE_MIOPEN": "OFF",
    "USE_MKL": "OFF",
    "USE_MSVC_MT": "OFF",
    "USE_NNPACK": "OFF",
    "USE_OPENCL": "OFF",
    "USE_OPENCL_GTEST": "/path/to/opencl/gtest",
    "USE_OPENMP": "none",
    "USE_PAPI": "OFF",
    "USE_PROFILER": "ON",
    "USE_PT_TVMDSOOP": "OFF",
    "USE_RANDOM": "ON",
    "USE_RELAY_DEBUG": "OFF",
    "USE_ROCBLAS": "OFF",
    "USE_ROCM": "OFF",
    "USE_RPC": "ON",
    "USE_RTTI": "ON",
    "USE_RUST_EXT": "OFF",
    "USE_SORT": "ON",
    "USE_SPIRV_KHR_INTEGER_DOT_PRODUCT": "OFF",
    "USE_STACKVM_RUNTIME": "OFF",
    "USE_TARGET_ONNX": "OFF",
    "USE_TENSORFLOW_PATH": "none",
    "USE_TENSORRT_CODEGEN": "OFF",
    "USE_TENSORRT_RUNTIME": "OFF",
    "USE_TFLITE": "OFF",
    "USE_TF_TVMDSOOP": "OFF",
    "USE_THREADS": "ON",
    "USE_THRUST": "OFF",
    "USE_VITIS_AI": "OFF",
    "USE_VULKAN": "OFF"
  }
```

junrushao committed:
[MetaSchedule][Runtime] Enhance Runner RandomFill (#11758)

ZihengJiang committed:
[COMMUNITY] Denise Kutnick -> Reviewer (#11778)

vinx13 committed:
[TIR, analysis] Add GetAutoTensorizeMappingInfo to generate transforms for auto tensorization (#11740)

This PR added a utility function `GetAutoTensorizeMappingInfo` to propose mapping from workload block iters to the iters in the tensor intrin. An example usage is conv2d, where the computation block has more iters than the matmul tensor intrin.

lhutton1 committed:
[TVMC] Fix tvmc run when using rpc (#11757)

* [TVMC] Fix tvmc run when using rpc

As described in #11707, the RPC mechanism does not support
objects of type Map which breaks the use of tvmc run when using
RPC after #9889. This commit intends to workaround this issue by
providing a fallback to the old implementation when RPC is being
used. Further, a test has been provided to help prevent this
regression in the future.

Change-Id: I70c1863d00098270e27c08ba834a3587e9132d69

* fix lint

Change-Id: I958cf4e19988d047bdd2e02f6475b9f70afe80c8

driazati committed:
[ci][docker] Remove Docker image upload prefix (#11769)

These should go to the same tag as in `tlcpack` so the only difference is the user

junrushao committed:
[MetaSchedule] Enhance parsing in JSONDatabase (#11791)

Originally, when failed with `std::stoi` and `std::stod`, the parser
disruptly stops and throw an incomprehensible error message, for
example, "stoi". This PR improves the user experience by detailing which
string causes the parsing issue.

A minor fix: out-of-range integers in parsing will now automatically fall
back to floating point numbers.

zxybazh committed:
[MetaSchedule][Minor] Update CPU Flush ArgParse Type (#11792)

Previously `cpu-flush` option existed as a boolean or integer argument, which is a bit counter-intuitive because for argparse, any non-empty string such as `False` will be parsed to `True` when using as a boolean and integer a little bit vague here IMHO. This PR used a function from `distutils` to directly parse input string to boolean, which makes the usage more stragiht-forward like `--cpu-flush True` or `--cpu-flush False`. Meanwhile it still supports usage of `0/1` and made sure the argument is always required.

chengven027 committed:
add layerNormal infer layout (#11784)

huajsj committed:
[docs] Fix incorrect command (#11630)

build/html get moved into _docs, update related document.

UlrikHjort committed:
[microTVM] [Fix] reboot include for Zephyr version >=2.6.0 (#11790)

* Fix reboot include for Zephyr version >=2.6.0

mehrdadh committed:
Fix apt install (#11781)

trahman-quic committed:
[HEXAGON] Slice ops added - add, subtract, multiply (#11529)

* [UPSTREAM][HEXAGON] Slice ops added - add, subtract, multiply

* Change to v68

* Change transform_numpy function call

* Do not disbale pylint errors and fix them

* Fix variable names

* Move the test file to topi

* Resolve conflict

* Modify init

Lunderberg committed:
[Arith] Simplification of ceil, log2, and left_shift (#11646)

* [TIR] Simplify expressions using tir.ceil and tir.log2

These expressions are introduced in `topi.math.ceil_log2`, and can
otherwise be propagated through to the generated kernel.

* [Arith] Added left shift handling to ConstIntBoundsAnalyzer

Previously, only right shift was handled.  These left shifts are
used in the `cuda.sort` implementation.

* Update to avoid left shift of negative numbers

* Updated rewriting of log2(x) to only occur in ceil(log2(x))

Per @wrongtest's request, to avoid rounding differences between
different devices.

* Avoid assumptions made of negative arguments to left-shift

* Recognize bounds of int(ceil(log2(arg)))

alter-xp committed:
[CI Image] support CSI-NN2 in ci_qemu (#11689)

* [CI Image] support CSI-NN2 in ci_qemu

* build CSI-NN2, download related toolchain and qemu

* using fixed csinn2 branch

lazycal committed:
[TE Schedule] Fix broken 2D softmax TE schedules when axis=0 (#11803)

* Support arbitrary reduce axis in softmax schedule.

* Fix lint.

masahi committed:
[CI] Update GPU image to add DNNL (#11786)

Requested by https://github.com/apache/tvm/issues/11774
Validated in https://ci.tlcpack.ai/blue/organizations/jenkins/tvm/detail/ci-docker-staging/260/

Commit:
[LLVM] Retrieve entire target string from LLVMModule (#11802)

The blob-embedding code creates a new LLVM module for which is needs more
information than just the target triple. The `_get_target_triple` function
in LLVMModule returned the triple with additional options appended to the
string. Instead of piggy-backing those extra options on top of the triple,
replace `_get_target_triple` with `_get_target_string`, which will return
the entire target string.

junrushao committed:
[TIR] Add SHash and SEqual to IndexMap (#11798)

guberti committed:
[skip ci] Use Stanford Cars mirror to fix CI docs build (#11812)

Commit:
Change new concat (#11800)

* changed x86/concat to use lists of ints instead of te.tensor.Tensor for loop extents and array offsets

* typos fixed

* removed unused import

* fixed micro model test

* fixed micro model test

jinhongyii committed:
[Relay][Op] MetaSchedule layout in TypeRel (#11819)

Co-authored-by: Junru Shao <junrushao1994@gmail.com>

dchauhan-arm committed:
[TFLite] Support quantized EQUAL op in TFLite frontend (#11520)

* [TFLite] Support quantized EQUAL op in TFLite frontend

Support EQUAL quantization operation conversion as part of issue #9187

* [TFLite] Support quantized EQUAL op in TFLite frontend

Update elementwise quantized test for EQUAL op
Change-Id: I3897d1ac07051ebfc10356ad45397117b592f878

lhutton1 committed:
[CPP-RPC] Fix command line argument capture (#11801)

There are a couple of instances where command line options use a "-",
however when capturing these values a "_" is used, meaning they
don't get captured and the default value is used instead. Fixing
this by renaming instances of "_" -> "-".

Change-Id: I9e083e25c5cc273298cd15df85a5862ee5f6722c

mehrdadh committed:
[microTVM][RVM] Reuse QEMU installation config and fix bug in RVM testing (#11808)

* refactor
* add llvm installation
* fix testing

yelite committed:
Merge branch 'main' into tunip

gbonik commented:
Will create a PR in Lite's fork instead
